Switch Kafka image from bitnami to confluentinc

gherceg opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi-dev.atlassian.net/browse/SAAS-13630

After encountering issues on M1 macs with the bitnami kafka image due to a lack of support, we are switching to an image that supports both arm and x86 architectures. The image release versioning is a bit odd, but can confirm that [v7.2.2 contains kafka version 3.2.2](https://docs.confluent.io/platform/current/release-notes/changelog.html#ak).

The config is slightly more involved with this image, so the changes we made are:


- `KAFKA_ADVERTISED_PORT` is deprecated in favor of `KAFKA_ADVERTISED_LISTENERS` which expects the form {LISTENER}://{HOST}:{PORT}
- The default is to use the security protocol (`PLAINTEXT` in this case) as the listener name (e.g., `PLAINTEXT://0.0.0.0:9092`), but to be explicit, we decided to set the listener name ourselves which is what `KAFKA_INTER_BROKER_LISTENER_NAME` specifies. 
- `KAFKA_LISTENER_SECURITY_PROTOCOL_MAP` explicitly sets the security protocol used by the listener.

Here are the notable config values that changed.

Before
```
advertised.listeners = PLAINTEXT://:9092
inter.broker.listener.name = null
listener.security.protocol.map = PLAINTEXT:PLAINTEXT,SSL:SSL,SASL_PLAINTEXT:SASL_PLAINTEXT,SASL_SSL:SASL_SSL
listeners = PLAINTEXT://:9092
```
After
```
advertised.listeners = HQ_KAFKA://localhost:9092
inter.broker.listener.name = HQ_KAFKA
listener.security.protocol.map = HQ_KAFKA:PLAINTEXT
listeners = HQ_KAFKA://0.0.0.0:9092
```

Relevant resources:
https://www.confluent.io/blog/kafka-listeners-explained/
https://kafka.apache.org/documentation/
## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Dev setup only. Tested locally and on an M1 machine, works in both scenarios.
### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

Reverting this PR will break local dev environments that run on M1 chips.

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
use ubuntu kafka image instead of bitnami

gherceg committed:
use confluentic kafka image instead of bitnami

gherceg committed:
remove deprecated KAFKA_ADVERTISED_HOST_NAME

gherceg committed:
do not overwrite hostname passed by local env

dannyroberts commented on docker/hq-compose.yml:
```diff
@@ -117,13 +117,16 @@ services:
 
   kafka:
-      image: bitnami/kafka:3.2
+      image: confluentinc/cp-kafka:7.2.2
```
> The image release versioning is a bit odd, but can confirm that [v7.2.2 contains kafka version 3.2.2](https://docs.confluent.io/platform/current/release-notes/changelog.html#ak).

Can you note this here in a comment? Otherwise if I came across this line looking for our current kafka version I would be quite confused!

gherceg replied:
Good point. See 5fae93be355ae6f5a56bfa4b58fef4273d9a429d

skodde committed:
Kafka and ZooKeeper docker configurations cleanup

skodde committed:
Use KAFKA_HOSTNAME for KAFKA_ADVERTISED_LISTENERS

Defaults to 'localhost', will be set to 'kafka' when running in
GitHub Actions.

gherceg committed:
add kafka version in comment

gherceg reviewed (commented):
Thanks for the changes to get the github actions build passing @skodde!

gherceg commented:
Given that Marco made changes in this PR, we will need approval from an additional reviewer. @snopoke if you get a chance to look at this Monday morning your time and approve, feel free to merge it as well.

dannyroberts reviewed (approved):
Reviewed with ?w=1 (i.e. https://github.com/dimagi/commcare-hq/pull/32275/files?w=1) to see a diff that ignores whitespace changes. While I didn't try to follow the logic for every env var change, the fact that the tests pass in github actions and it satisfied the testing you did locally (and my impression that nothing here looks alarming) is enough for me!

skodde commented on docker/hq-compose.yml:
```diff
@@ -111,32 +111,33 @@ services:
     image: zookeeper:3.7
-    ports:
-      - "2181:2181"
+    expose:
+      - "2181"
```
Yes to the first part: `expose` maps to the `EXPOSE` Docker instruction, which is supposed to be mostly informative, as containers on the same network, with the default bridge network, should be able to connect to each other regardless of this instruction. `ports` instead maps to `docker run -p` (or equivalent implementation) which publishes the port(s) to the host.
We want to have both as we want to be able to connect to these containers also from the host, and I only moved the `ports` attribute to `docker/hq-compose-services.yml` instead, as this seems to be the structure we've been using for all other containers.

References:
https://docs.docker.com/compose/compose-file/#expose
https://docs.docker.com/compose/compose-file/#ports
https://docs.docker.com/engine/reference/builder/#expose
https://docs.docker.com/network/#network-drivers

gherceg commented:
@dimagi/dimagi-dev the previous kafka docker image had issues running on M1 macs. This new image supports both architectures. Run the following on an updated checkout of master to update your local kafka container:
```
./scripts/docker rm -sv kafka zookeeper
./scripts/docker up -d kafka zookeeper
```
